Add Float 2d snap rounding (#8797)
CGAL/cgal: Snap_rounding_2/doc/Snap_rounding_2/Snap_rounding_2.txt

\section Vertical_slab_snap_rounding Vertical Slab Snap Rounding Algorithm

Vertical Slab Snap Rounding algorithm is inspired by 3D snap rounding techniques \cgalCite{valque2024snaprounding}. It subdivides the segments vertical slabs to prevent intersections during the rounding process.
The algorithm is illustrated in Figure \cgalFigureRef{fig_float_snap} and proceeds as follows. First, the segments are subdivided
at their (potential) intersection points. Then the vertices are processed from left to right. Whenever
a vertex lies too close to a segment, that segment is subdivided at the \f$x\f$-coordinate of the vertex.
Finally, the vertices are rounded.

\cgalFigureBegin{fig_float_snap,float_snap_fig.svg}
An arrangement of segments before subdividing the segments (a), after subdividing the segments (b) and after rounding (c).
\cgalFigureEnd

\section FSR_Perf Snap Rounding Performance
We evaluate the different methods on two tests:
In the first test, the segment endpoints are chosen uniformly and independantly in the unit box.
In the second test, we generate perturbed copies of a given segment. This experiment is repeated several times using different input segments.

> The two pixel sizes for hot pixel snap rounding are chosen such that the rounding shift is of the same order of magnitude as the rounding induced by using `double` or `float` coordinates.

Reviewer: ```suggestion
The two pixel sizes for Hot Pixel Snap Rounding are chosen such that the rounding shift is of the same order of magnitude as the rounding induced by using `double` or `float` coordinates.
```

The first test is standard but typically exhibits few, if any, rounding issues.
The second test is deliberately designed to be difficult, as it produces a large number of rounding issues when rounding is performed naively.

> In these settings, vertical slab snap rounding significantly outperforms hot pixel snap rounding and remains of the same order of magnitude than computing the intersections without snap.

Reviewer: ```suggestion
In these settings, vertical slab snap rounding significantly outperforms Hot Pixel Snap Rounding and remains of the same order of magnitude than computing the intersections without snap.
```

The result of our test is summarized in the next table:

<center>

> | Test Case | Computing intersection<br> (Unsafe rounding) | Vertical slab SR <br> (double) | Vertical Slab SR <br> (float)  | Hot pixel SR <br>(pixel size \f$10^{-15}\f$ ) | Hot pixel SR <br>(pixel size \f$10^{-7}\f$) |

Reviewer: ```suggestion
| Test Case | Computing intersection<br> (Unsafe rounding) | Vertical Slab SR <br> (double) | Vertical Slab SR <br> (float)  | Hot Pixel SR <br>(pixel size \f$10^{-15}\f$ ) | Hot Pixel SR <br>(pixel size \f$10^{-7}\f$) |
```

|-----------|-------|-------|-------|-------|-------|
| 500 Random Segments | 0.066s | 0.213s | 0.172s | 6.14s | 3.81s |
| 1000 Random Segments | 0.241s | 0.928s | 0.769s | 52.1s | 53.5s |
| 1500 Random Segments | 0.566s | 2.88s | 2.93s | 201s | 197s |
| 2000 Random Segments | 1.15s | 5.37s | 5.468s | 452s | 446s |
| 200 Almost identical Segments | 0.073s | 0.114s | 11.9s | 2.13s | 3.41s |
| 300 Almost identical Segments | 0.133s | 0.284s | 44.1s | 8.58s | 14.2s |
| 400 Almost identical Segments | 0.262s | 0.620s | 102s | 32.2s | 42.0s |
| 500 Almost identical Segments | 0.404s | 1.31s | 206s | 69.2s | 87s |
</center>
\subsection FSR_comp Snap Rounding Complexity
Both methods run in \f$O(n \log n + k)\f$, where \f$n\f$ is the number of input segments
and \f$k\f$ is the number of vertices created during the process. In the worst case, the number of created
vertices is \f$O(n^3)\f$, or \f$O(n^2)\f$ if the input segments are free of intersections.

\section Snap_rounding_2API Snap Rounding API and Examples

> Both algorithms can be invoked through the \ref snap_rounding_2_fct "CGAL::snap_rounding_2()" function.
> \ref snap_rounding_2_fct "CGAL::snap_rounding_2()" take a range of input segments and, by default, produces a range of polylines, each polyline corresponding to an input segment.

Reviewer: ```suggestion
Both algorithms can be invoked through the \ref snap_rounding_2_fct "CGAL::snap_rounding_2()" function, which takes a range of input segments and, by default, produces a range of polylines, each polyline corresponding to an input segment.
```

Consequently, duplicate segments may appear in the output, for instance when multiple input segments collapse. When the parameter `output_unique_segments` is set to `true`,
the polylines are decomposed into individual segments (they are represented as polylines with two points), and duplicates are removed.
An overload also supports as input and output polygon ranges.

> By default, the Vertical Slab Snap Rounding algorithm to double precision is used. The other method or other rounding schemes can be used depending on the traits class provided.

Reviewer: ```suggestion
By default, the Vertical Slab Snap Rounding algorithm to double precision is used. Other methods or other rounding schemes can be used depending on the traits class provided.
```

Specific methods also exist: `CGAL::vertical_slab_snap_rounding_2()` and `CGAL::hot_pixel_snap_rounding_2`.

\subsection Float_snap_rounding_Traits Snap Rounding Traits

> 
> 

Reviewer: ```suggestion

```

When using Hot Pixel SR, the traits class must be a model of `HotPixelSnapRoundingTraits_2`. The class `Hot_pixel_snap_rounding_traits_2` is such a model.

> When using Vertical Slab SR, the traits class must be a model of  `VerticalSlabSnapRoundingTraits_2`. By default,

Reviewer: ```suggestion
When using Vertical Slab SR, the traits class must be a model of `VerticalSlabSnapRoundingTraits_2`. By default,
```

> the traits class `Double_grid_snap_rounding_traits_2`, which rounds coordinates to double precision floating point,
> is used. The package also provides `Float_grid_snap_rounding_traits_2` and `Integer_grid_snap_rounding_traits_2`, which rounds coordinates to

Reviewer: ```suggestion
the traits class `CGAL::Double_grid_snap_rounding_traits_2`, which rounds coordinates to double precision floating point,
is used. The package also provides `CGAL::Float_grid_snap_rounding_traits_2` and `CGAL::Integer_grid_snap_rounding_traits_2`, which rounds coordinates to
```

single precision floating point and integers, respectively. Other traits classes can be used to round to different representations or rounding schemes as long as
the rounding operation preserves the order of vertices along the \f$x\f$ and \f$y\f$ directions.
See `VerticalSlabSnapRoundingTraits_2` for more details.

\subsection Snap_rounding_2Four Snap Rounding Examples

The following example snaps on integers using Hot Pixel SR a instance of four segments.

\cgalExample{Snap_rounding_2/snap_rounding_to_integer.cpp}

The following example snaps on double a non-trivial instance of five segments.

\cgalExample{Snap_rounding_2/snap_rounding_to_double.cpp}

The following example generates an Iterative SR representation
of an arrangement of four line segments and outputs the resulting polylines in a plane
tiled with one-unit square pixels.

\cgalExample{Snap_rounding_2/iterative_snap_rounding.cpp}

The package is supplied with a graphical demo program that opens a window,
allows the user to edit segments dynamically, applies a selected
snap-rounding procedures, and displays the result onto the same window
(see `<CGAL_ROOT>/GraphicsView/demo/Snap_rounding_2/Snap_rounding_2.cpp`).

> 
> 

Reviewer: ```suggestion

```
